### PR TITLE
build: Store the target_cpu in an "arch" file in the resulting ZIP

### PR DESCRIPTION
**Pull request by jviotti**
We have an application that interacts with many different versions of
the Electron.js ZIP artifacts at the same time.

The ZIP artifacts currently ship with a `version` text file that matches
the content of the `ELECTRON_VERSION` file present at the root of this
repository. The presence of this file makes it easy for an application
consuming the ZIP artifacts to introspect its contents with regards to
the version information.

This commit extends such pattern to include the value of the GN
`target_cpu` value into an `arch` file to also easily introspect on the
distributable ZIP architecture. For example, building Electron.js on an
Intel 64-bit host results in `arch` containing the string `x64`. This
commit stores the value of `target_cpu` instead of `host_cpu` to produce
the expected result when cross-compiling Electron.js.

Notes: Add an `arch` file to the distributable ZIP artifacts.
See: https://gn.googlesource.com/gn/+/refs/heads/main/docs/reference.md#var_target_cpu
Signed-off-by: Juan Cruz Viotti <jv@jviotti.com>

### CODE REVIEW

**Comment by jviotti** [comment]
If you agree with the improvement, let me know if there is a better way to set it up on GN than the `action` with the dummy Python script I included and I'll update the PR accordingly.

**Comment by jviotti** [comment]
The Linux builds seemed to have failed because the Goma proxy crashed.

**Review by zcbenz — COMMENTED** [review]
I prefer using a JSON file that can be extended to included more information in future.

I'm not sure which work group is the best to reference, but I would like to hear more opinions from @electron/wg-releases.

**Review by MarshallOfSound — COMMENTED** [review]
I'm like -0.5 on this change, not strongly against it but it feels strange adding this to the zip file and it opens us up to other questions like "what about `platform`" and "what about `x`".

The version we have the `version` file iirc is to specifically handle the case where it is unknown during package install time whether or not the file in the `dist` folder of `node_modules` is the correct version of Electron.  Refs: https://github.com/electron/electron/blob/master/npm/install.js#L36

Can you explain / give an example of a scenario where the architecture of a zip file (that you / your build scripts must have downloaded) is unknown? It seems like the kind of thing that could be solvable via correct usage of `@electron/get` rather than relying on the contents of the zip file itself.

**Comment by jviotti** [comment]
@MarshallOfSound 

> Can you explain / give an example of a scenario where the architecture of a zip file (that you / your build scripts must have downloaded) is unknown? It seems like the kind of thing that could be solvable via correct usage of @electron/get rather than relying on the contents of the zip file itself.

We ([Postman](https://www.postman.com)) are maintaining a desktop framework build around a fork of Electron + other goodies for our app. This desktop framework comes with a CLI tool for our team members to work on the Electron application. One of the commands that we support is the `install` command, which essentially wraps a particular version of NPM (so that we all run the exact same thing) and makes sure that `npm install` is ran with the correct configuration to always produce working native add-ons (it forces them to be built from source using the development environment we provide, sets the right Node.js ABI, target architecture as we support certain cross-compile scenarios, etc).

When being ran, the CLI that we provide will try to find the corresponding Electron artifact from our fork, download it, cache it, and resolve the path to it (similar to `@electron/get`). The Electron fork that we maintain has various Postman-specific changes and is therefore versioned independently of the main Electron.js project. When setting the `npm_config_target` option during NPM install, we introspect into the `version` file to resolve the right Electron.js version that is used by the particular version of our fork.

In the same way, our `install` command has to tell NPM the architecture that it has to build the native add-ons for (i.e. `npm_config_target_arch`). The `install` command is not necessarily the one that causes the artifacts to be downloaded & cached. It might have been another command on a previous execution of the command-line tool. At that point, `install` knows where the artifact is located and the Electron artifact that it should use, but the target architecture to build on (i.e. depending on the cross-compilation settings that the engineer requested) depends on the artifact that the command was told to use. 

We explored a couple of approaches:

- We can unzip the artifact, locate the path to the binary, and do `ELECTRON_RUN_AS_NODE=1 path/to/electron -p 'process.arch'`. This is what we were doing before, but required us to unzip the artifact at the dependencies installation time instead of at the packaging step and actually have to run it instead of simply i.e. reading a text file from the zip without even unzipping the entire contents
- We can store the desired target architecture configuration in a way that the `install` command can retrieve it and thus have both the path to the artifact and the target architecture. The reason why we didn't continue doing this was that we rather kept the `install` command simple, internally take a path to an Electron ZIP and just know what to do with it without requiring additional information that could get out of sync (essentially it will just take the ZIP as the sole source of truth, which felt like an slightly better design)

The presence of the `version` file in the Electron.js artifacts gave us the idea than a similar `arch` file would actually give us both of the things mentioned above: we can inspect the zip without extracting and running the whole thing, and we can still keep the `install` internal command taking the ZIP as the sole source of truth. We have been doing this since then in our fork.

In summary, we have other approaches to do this (and we are maintaining our fork anyway). However, we are already introspecting on the `version` file to keep various of our commands dumb (they just take the ZIP they must use and figure everything out from there without risking getting into conflicting parameters), so doing the same thing with the `arch` felt natural. 

@zcbenz 

> I prefer using a JSON file that can be extended to included more information in future.

That's also a nice approach. We're happy to help if that's the path forward you prefer.

**Comment by jviotti** [comment]
Self-identifying artifacts might be an interesting long-term feature IMHO. Right now, things like the OS and the architecture are present in the filename of the ZIP artifact. If for some reason you unzip with the default output and remove the original ZIP, then you have no easy way to tell what the artifacts was. Even then, you might have to rely on parsing the filename to extract out the architecture, which feels fragile.

Surely you can rely on certain heuristics (like whether there is an `Electron.app` directory) or even attempt to execute the binary itself as node, but there are still cases where that might not be enough. For example, if I'm running macOS and have a Linux artifact, then you might not be able to know the architecture of it unless you maybe inspect the binaries with a macOS ELF-friendly tool? 

Looking at it with such lense, having the artifact show enough information about itself in a generic way (i.e. plain text or JSON) so that you can understand what it is even if you are not running that same operating system feels like a good design.

**Comment by jviotti** [comment]
In any case, this sounds like a big topic for discussion, so I'm also happy to retire the PR until we find a way forward (or decide to not focus on it) 😄

**Comment by nornagon** [comment]
Hm... I don't see why you wouldn't already know the target, if you're fetching the build with @electron/get or similar.

**Comment by zcbenz** [comment]
The discussion has halted for a while, and we have 2 maintainers opposing this change and none supporting it, so I don't think we can push this forward and I'm closing this PR.
